src/serac/numerics/functional/domain.cpp

```diff
@@ -0,0 +1,481 @@
+  mesh.GetVertices(vertices);
+
+  mfem::Array<int> edge_id_to_bdr_id;
+  if (d == 2) {
+    edge_id_to_bdr_id = mesh.GetFaceToBdrElMap();
+  }
+
+  int num_edges    = mesh.GetNEdges();
+  int num_vertices = mesh.GetNV();
+  for (int i = 0; i < num_edges; i++) {
+    mfem::Array<int> vertex_ids;
+    mesh.GetEdgeVertices(i, vertex_ids);
+
+    std::vector<tensor<double, d>> x(2);
+    for (int v = 0; v < 2; v++) {
+      for (int j = 0; j < d; j++) {
+        x[uint32_t(v)][j] = vertices[j * num_vertices + vertex_ids[v]];
+      }
+    }
+
+    if constexpr (d == 2) {
+      int bdr_id = edge_id_to_bdr_id[i];
+      int attr   = (bdr_id > 0) ? mesh.GetBdrAttribute(bdr_id) : -1;
+      if (predicate(x, attr)) {
+        output.edges_.push_back(i);
+      }
+    } else {
+      if (predicate(x)) {
+        output.edges_.push_back(i);
+      }
+    }
+  }
+
+  return output;
+}
+
+Domain Domain::ofEdges(const mfem::Mesh& mesh, std::function<bool(std::vector<vec2>, int)> func)
+{
+  return domain_of_edges<2>(mesh, func);
+}
+
+Domain Domain::ofEdges(const mfem::Mesh& mesh, std::function<bool(std::vector<vec3>)> func)
+{
+  return domain_of_edges<3>(mesh, func);
+}
+
+///////////////////////////////////////////////////////////////////////////////////////
+///////////////////////////////////////////////////////////////////////////////////////
+
+template <int d>
+static Domain domain_of_faces(const mfem::Mesh&                                        mesh,
+                              std::function<bool(std::vector<tensor<double, d>>, int)> predicate)
+{
+  assert(mesh.SpaceDimension() == d);
+
+  Domain output{mesh, 2 /* faces are 2-dimensional */};
+
+  // layout is undocumented, but it seems to be
+  // [x1, x2, x3, ..., y1, y2, y3 ..., (z1, z2, z3, ...)]
+  mfem::Vector vertices;
+  mesh.GetVertices(vertices);
+
+  mfem::Array<int> face_id_to_bdr_id;
+  if (d == 3) {
+    face_id_to_bdr_id = mesh.GetFaceToBdrElMap();
+  }
+
+  // faces that satisfy the predicate are added to the domain
+  int num_faces;
+  if (d == 2) {
+    num_faces = mesh.GetNE();
+  } else {
+    num_faces = mesh.GetNumFaces();
+  }
+
+  int tri_id  = 0;
+  int quad_id = 0;
+
+  int num_vertices = mesh.GetNV();
+  for (int i = 0; i < num_faces; i++) {
+    mfem::Array<int> vertex_ids;
+
+    if (mesh.Dimension() == 2) {
+      mesh.GetElementVertices(i, vertex_ids);
+    } else {
+      mesh.GetFaceVertices(i, vertex_ids);
+    }
+
+    std::vector<tensor<double, d>> x(std::size_t(vertex_ids.Size()));
+    for (int v = 0; v < vertex_ids.Size(); v++) {
+      for (int j = 0; j < d; j++) {
+        x[uint32_t(v)][j] = vertices[j * num_vertices + vertex_ids[v]];
+      }
+    }
+
+    int attr;
+    if (d == 2) {
+      attr = mesh.GetAttribute(i);
+    } else {
+      int bdr_id = face_id_to_bdr_id[i];
+      attr       = (bdr_id > 0) ? mesh.GetBdrAttribute(bdr_id) : -1;
+    }
+
+    if (predicate(x, attr)) {
+      if (x.size() == 3) {
+        output.tris_.push_back(tri_id);
```

Comment:
I'm a bit confused how this works. The vertices and the edges store the MFEM index, but this just seems to be a list of 1...N for how many faces are in the domain. Am I missing some logic here?

Reply:
mfem::Mesh puts all of its elements (of any geometry) into a single list, but mfem's "E-vector" data layout only works with elements of one geometry. So, I'm splitting up mfem's element indices into separate lists for each element geometry. For example:

| mfem::Geometry | mfem element id | tri id | quad id |
| -------------- | --------------- | ------ | ------- |
| Triangle       | 0               | 0      |         |
| Triangle       | 1               | 1      |         |
| Square         | 2               |        | 0       |
| Triangle       | 3               | 2      |         |
| Square         | 4               |        | 1       |
| Square         | 5               |        | 2       |
| Square         | 6               |        | 3       |

Then, if an element satisfes the predicate it is added to the `Domain`.

Vertex and edge indices are treated slightly differently since there is only 1 kind of "element" for 0D and 1D, so no separation is required.

Reply:
Ahhh I missed that the quad/tri ids were always being incremented even if they didn't satisfy the predicate. Thank you for the explanation!